src/formats/CML.cpp

```diff
@@ -0,0 +1,581 @@
+        return;
+    }
+
+    read_property_(container, node, title.as_string());
+}
+
+void CMLFormat::read(Frame& frame) {
+    const auto& current = *current_;
+    for (auto attribute : current.attributes()) {
+        auto name = std::string(attribute.name());
+        if (name == "id") { // do nothing, required junk for validation
+        } else if (name == "xmlns") {
+        } else if (name == "xmlns:cml") {
+        } else if (name == "xmlns:units") {
+        } else if (name == "xmlns:xsd") {
+        } else if (name == "xmlns:iupac") {
+        } else if (name == "title") {
+            frame.set("title", attribute.as_string());
+        } else {
+            warning("[CML] Unknown molecule attribute: " + name);
+        }
+    }
+
+    auto name_node = current.child("name");
+    if (name_node) {
+        frame.set("name", name_node.text().as_string());
+    }
+
+    auto crystal_node = current.child("crystal");
+    if (crystal_node) {
+        double a, b, c, alpha, beta, gamma;
+        a = b = c = alpha = beta = gamma = 0.0;
+        for (auto cell_param : crystal_node.children("scalar")) {
+            auto title = cell_param.attribute("title");
+            if (title) {
+                auto name = std::string(title.value());
+                if (name == "a") {
+                    a = cell_param.text().as_double();
+                } else if (name == "b") {
+                    b = cell_param.text().as_double();
+                } else if (name == "c") {
+                    c = cell_param.text().as_double();
+                } else if (name == "alpha") {
+                    alpha = cell_param.text().as_double();
+                } else if (name == "beta") {
+                    beta = cell_param.text().as_double();
+                } else if (name == "gamma") {
+                    gamma = cell_param.text().as_double();
+                } else {
+                    warning("[CML] Unknown crytal scalar: " + name);
+                }
+            }
+        }
+        frame.set_cell(UnitCell(a, b, c, alpha, beta, gamma));
+    }
+
+    // Frame level properties
+    auto properties_node = current.child("propertyList");
+    if (properties_node) {
+        for (auto property : properties_node.children("property")) {
+            auto title = property.attribute("title");
+            if (!title) {
+                warning("[CML] Skipping untitled property");
+                continue;
+            }
+
+            auto title_str = title.as_string();
+            auto scalar_node = property.child("scalar");
+            if (!scalar_node) {
+                warning("[CML] {} has no scalar associated with it!", title_str);
+                continue;
+            }
+
+            read_property_(frame, scalar_node, title_str);
+        }
+    }
+
+    // Read the atoms
+    std::unordered_map<std::string, size_t> ref_to_id;
+
+    // TODO: Read string vector versions instead of just the nodes
```

Comment:
Is this still TODO ?

Reply:
Yes, I could only find one program that created CML files in this manner and I had to modify its settings quite a bit. Other CML parsers (Avogadro, chemkit) ignore this method in any case.